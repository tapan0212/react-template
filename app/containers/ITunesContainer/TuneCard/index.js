/**
 *
 * TuneCard
 *
 */

import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage as T } from 'react-intl';
import { Row, Col } from 'antd';
import { PlayCircleOutlined, PauseCircleOutlined, InfoCircleOutlined } from '@ant-design/icons';
import { useHistory } from 'react-router-dom';
import { colors } from '@themes';
import If from '@app/components/If';

const CustomTuneCard = styled.div`
  && {
    margin: 2rem 0;
    height: 10rem;
    padding: 1rem;
    border: 2px solid ${colors.primary};
    cursor: pointer;
  }
`;

const CustomPlayButton = styled(PlayCircleOutlined)`
  float: right;
  font-size: 2rem;
  margin-top: 3rem;
  margin-right: 3rem;
  cursor: pointer;
`;
const CustomPauseButton = styled(PauseCircleOutlined)`
  float: right;
  font-size: 2rem;
  margin-top: 3rem;
  margin-right: 3rem;
  cursor: pointer;
`;
const CustomInfoButton = styled(InfoCircleOutlined)`
  float: right;
  font-size: 2rem;
  margin-top: 3rem;
  margin-right: 5rem;
  cursor: pointer;
`;
function TuneCard({ song, currentTune, dispatchCurrentTune, dispatchSelectedTune }) {
  const tuneRef = useRef(null);
  const history = useHistory();

  useEffect(() => {
    (() =>
      currentTune && song && currentTune?.trackId === song?.trackId
        ? tuneRef.current.play()
        : tuneRef?.current?.pause())();
  }, [currentTune, song]);

  const artistName = song.artistName;
  const trackName = song.trackName;
  const release = song.releaseDate.substring(0, 4);
  return (
    <CustomTuneCard data-testid="tune-card">
      <Row>
        <Col span={12}>
          <div style={{ marginTop: '1.5rem' }}>
            <T id="track_name" values={{ trackName }} />
          </div>
          <div>
            <T id="artist_name" values={{ artistName }} />
          </div>
          <div>
            <T id="released" values={{ release }} />
          </div>
        </Col>
        <Col span={12}>
          <If condition={currentTune?.trackId !== song?.trackId}>
            <CustomPlayButton
              data-testid="btn1"
              onClick={e => {
                dispatchCurrentTune(song);
              }}
            />
          </If>
          <If condition={currentTune?.trackId === song?.trackId}>
            <CustomPauseButton
              data-testid="btn2"
              onClick={e => {
                dispatchCurrentTune(null);
              }}
            />
          </If>

          <CustomInfoButton
            data-testid="info_btn"
            onClick={() => {
              dispatchSelectedTune(song);
              history.push(`/tune/${song.trackId}`);
            }}
          />
        </Col>
      </Row>
      <audio ref={tuneRef}>
        <source src={song?.previewUrl}></source>
      </audio>
    </CustomTuneCard>
  );
}

TuneCard.propTypes = {
  song: PropTypes.object,
  currentTune: PropTypes.object,
  dispatchCurrentTune: PropTypes.func,
  dispatchSelectedTune: PropTypes.func
};

export default TuneCard;
