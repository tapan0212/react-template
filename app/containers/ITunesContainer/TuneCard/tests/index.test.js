/**
 *
 * Tests for TuneCard
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl } from '@utils/testUtils';
import TuneCard from '../index';

describe('<TuneCard />', () => {
  let dispatchCurrentTuneSpy;
  const mockSongData = {
    trackId: 1,
    artistName: 'Ed',
    trackName: 'Perfect',
    releaseDate: '2017-05-10'
  };
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TuneCard song={mockSongData} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should dispatch dispatchCurrentTune when clicked on button', () => {
    dispatchCurrentTuneSpy = jest.fn();
    const { getByTestId } = renderWithIntl(
      <TuneCard dispatchCurrentTune={dispatchCurrentTuneSpy} song={mockSongData} />
    );
    fireEvent.click(getByTestId('btn1'));
    expect(dispatchCurrentTuneSpy).toBeCalled();
  });

  // it('should dispatch dispatchSelectedTune when clicked on info btn', () => {
  //   dispatchSelectedTuneSpy = jest.fn();
  //   const { getByTestId } = renderWithIntl(
  //     <TuneCard dispatchSelectedTune={dispatchSelectedTuneSpy} history={pushSpy} song={mockSongData} />
  //   );
  //   fireEvent.click(getByTestId('info_btn'));
  //   expect(dispatchSelectedTuneSpy).toBeCalled();
  // })

  it('should show play button', () => {
    const { getByTestId } = renderWithIntl(
      <TuneCard song={mockSongData} currentTune={{ trackId: 2 }} dispatchCurrentTune={dispatchCurrentTuneSpy} />
    );
    expect(getByTestId('btn1')).toBeTruthy();
  });
  it('should show pause button', () => {
    const { getByTestId } = renderWithIntl(
      <TuneCard song={mockSongData} currentTune={{ trackId: 1 }} dispatchCurrentTune={dispatchCurrentTuneSpy} />
    );
    expect(getByTestId('btn2')).toBeTruthy();
  });
});
