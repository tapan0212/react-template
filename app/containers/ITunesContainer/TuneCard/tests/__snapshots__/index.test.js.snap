// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TuneCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="TuneCard__CustomTuneCard-kiui55-0 gZBdDs"
      data-testid="tune-card"
    >
      <div
        class="ant-row"
      >
        <div
          class="ant-col ant-col-12"
        >
          <div
            style="margin-top: 1.5rem;"
          >
            Tune: Perfect
          </div>
          <div>
            Artist Name: Ed
          </div>
          <div>
            Release Year: 2017
          </div>
        </div>
        <div
          class="ant-col ant-col-12"
        >
          <span
            aria-label="play-circle"
            class="anticon anticon-play-circle TuneCard__CustomPlayButton-kiui55-1 cuoJgz"
            data-testid="btn1"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="play-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M719.4 499.1l-296.1-215A15.9 15.9 0 00398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 000-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
              />
            </svg>
          </span>
          <span
            aria-label="info-circle"
            class="anticon anticon-info-circle TuneCard__CustomInfoButton-kiui55-3 cUuXlk"
            data-testid="info_btn"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="info-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
              />
            </svg>
          </span>
        </div>
      </div>
      <audio>
        <source />
      </audio>
    </div>
  </div>
</body>
`;
